# 1. Transforming Data with dplyr
Learn verbs you can use to transform your data, including select, filter, arrange, and mutate. You'll use these functions to modify the counties dataset to view particular observations and answer questions about the data

## The counties dataset
This particular dataset is from the 2015 United States Census.

### Chapter 1 verbs
- `select()`
- `filter()`
- `arrange()`
- `mutate()`

In [1]:
library(dplyr)

Warning message:
"package 'dplyr' was built under R version 3.6.3"
Attaching package: 'dplyr'

The following objects are masked from 'package:stats':

    filter, lag

The following objects are masked from 'package:base':

    intersect, setdiff, setequal, union



In [3]:
counties <- readRDS("counties.rds")

In [4]:
head(counties)

census_id state   county  region metro    population men   women hispanic
1 1001      Alabama Autauga South  Metro     55221     26745 28476 2.6     
2 1003      Alabama Baldwin South  Metro    195121     95314 99807 4.5     
3 1005      Alabama Barbour South  Nonmetro  26932     14497 12435 4.6     
4 1007      Alabama Bibb    South  Metro     22604     12073 10531 2.2     
5 1009      Alabama Blount  South  Metro     57710     28512 29198 8.6     
6 1011      Alabama Bullock South  Nonmetro  10678      5660  5018 4.4     
  white ... other_transp work_at_home mean_commute employed private_work
1 75.8  ... 1.3          1.8          26.5         23986    73.6        
2 83.1  ... 1.4          3.9          26.4         85953    81.5        
3 46.2  ... 1.5          1.6          24.1          8597    71.8        
4 74.5  ... 1.5          0.7          28.8          8294    76.8        
5 87.9  ... 0.4          2.3          34.9         22189    82.0        
6 22.2  ... 1.7          2.8          27.5          3865    79.5        
  public_work self_employed family_work unemployment land_area
1 20.9        5.5           0.0          7.6          594.44  
2 12.3        5.8           0.4          7.5         1589.78  
3 20.8        7.3           0.1         17.6          884.88  
4 16.1        6.7           0.4          8.3          622.58  
5 13.5        4.2           0.4          7.7          644.78  
6 15.1        5.4           0.0         18.0          622.81

- Select the following four columns from the `counties` variable:
`state`, `county`, `population`, `poverty`

In [5]:
counties %>%
  # Select the columns
  select(state, county, population, poverty)

state     county      population poverty
1    Alabama   Autauga      55221     12.9   
2    Alabama   Baldwin     195121     13.4   
3    Alabama   Barbour      26932     26.7   
4    Alabama   Bibb         22604     16.8   
5    Alabama   Blount       57710     16.7   
6    Alabama   Bullock      10678     24.6   
7    Alabama   Butler       20354     25.4   
8    Alabama   Calhoun     116648     20.5   
9    Alabama   Chambers     34079     21.6   
10   Alabama   Cherokee     26008     19.2   
11   Alabama   Chilton      43819     19.1   
12   Alabama   Choctaw      13395     23.4   
13   Alabama   Clarke       25070     24.7   
14   Alabama   Clay         13537     16.7   
15   Alabama   Cleburne     15002     17.0   
16   Alabama   Coffee       50884     17.1   
17   Alabama   Colbert      54444     17.4   
18   Alabama   Conecuh      12865     33.8   
19   Alabama   Coosa        11027     19.3   
20   Alabama   Covington    37886     20.2   
21   Alabama   Crenshaw     13938     16.7   
22   Alabama   Cullman      80965     18.5   
23   Alabama   Dale         49866     19.1   
24   Alabama   Dallas       42154     35.0   
25   Alabama   DeKalb       71068     18.7   
26   Alabama   Elmore       80763     13.6   
27   Alabama   Escambia     37935     24.2   
28   Alabama   Etowah      103766     19.7   
29   Alabama   Fayette      16896     18.3   
30   Alabama   Franklin     31634     22.5   
...  ...       ...         ...        ...    
3109 Wisconsin Washburn     15700     13.1   
3110 Wisconsin Washington  132921      5.6   
3111 Wisconsin Waukesha    393873      5.3   
3112 Wisconsin Waupaca      52125     10.4   
3113 Wisconsin Waushara     24321     10.3   
3114 Wisconsin Winnebago   169004     12.0   
3115 Wisconsin Wood         74012     11.0   
3116 Wyoming   Albany       37565     26.9   
3117 Wyoming   Big Horn     11895     12.1   
3118 Wyoming   Campbell     48013      7.2   
3119 Wyoming   Carbon       15739     15.3   
3120 Wyoming   Converse     14101      9.4   
3121 Wyoming   Crook         7229      5.0   
3122 Wyoming   Fremont      40755     14.2   
3123 Wyoming   Goshen       13544     17.9   
3124 Wyoming   Hot Springs   4809     10.7   
3125 Wyoming   Johnson       8600      6.3   
3126 Wyoming   Laramie      95431     10.1   
3127 Wyoming   Lincoln      18316      7.9   
3128 Wyoming   Natrona      80011     10.7   
3129 Wyoming   Niobrara      2516     14.9   
3130 Wyoming   Park         28985      7.2   
3131 Wyoming   Platte        8750     16.1   
3132 Wyoming   Sheridan     29738      8.3   
3133 Wyoming   Sublette     10117      8.1   
3134 Wyoming   Sweetwater   44772     11.5   
3135 Wyoming   Teton        22311      7.4   
3136 Wyoming   Uinta        20930     13.7   
3137 Wyoming   Washakie      8400     13.8   
3138 Wyoming   Weston        7152     14.1

---
## The filter and arrange verbs


In [12]:
counties_selected_scpu <- counties %>%
    select(state, county, population, unemployment)

### Arrange

In [13]:
counties_selected_scpu %>%
    arrange(population)

state         county                   population unemployment
1    Hawaii        Kalawao                   85         0.0        
2    Texas         King                     267         5.1        
3    Nebraska      McPherson                433         0.9        
4    Montana       Petroleum                443         6.6        
5    Nebraska      Arthur                   448         4.0        
6    Nebraska      Loup                     548         0.7        
7    Nebraska      Blaine                   551         0.7        
8    New Mexico    Harding                  565         6.0        
9    Texas         Kenedy                   565         0.0        
10   Colorado      San Juan                 606        13.8        
11   Alaska        Yakutat City and Borough 643         7.9        
12   North Dakota  Slope                    673         0.0        
13   Nebraska      Thomas                   675         0.0        
14   Nebraska      Hooker                   681         2.9        
15   Texas         Borden                   705         0.7        
16   Nebraska      Keya Paha                711         1.1        
17   Colorado      Mineral                  733         1.6        
18   Montana       Golden Valley            756         3.5        
19   Nebraska      Grant                    769         3.5        
20   Utah          Daggett                  776         4.3        
21   Texas         McMullen                 778        14.1        
22   South Dakota  Jones                    781         3.6        
23   Montana       Treasure                 812         0.8        
24   Nebraska      Banner                   820         5.4        
25   Texas         Kent                     821         1.5        
26   Nebraska      Wheeler                  847         2.9        
27   Nebraska      Logan                    851         0.9        
28   Colorado      Hinsdale                 874         3.0        
29   Idaho         Clark                    901         3.9        
30   Texas         Terrell                  921         3.7        
...  ...           ...                      ...        ...         
3109 Ohio          Cuyahoga                  1263189   10.6        
3110 Florida       Hillsborough              1302884    9.0        
3111 New York      Nassau                    1354612    6.4        
3112 Florida       Palm Beach                1378806    9.5        
3113 New York      Bronx                     1428357   14.0        
3114 California    Sacramento                1465832   11.8        
3115 New York      Suffolk                   1501373    6.4        
3116 Pennsylvania  Philadelphia              1555072   13.9        
3117 Massachusetts Middlesex                 1556116    6.1        
3118 California    Alameda                   1584983    8.3        
3119 New York      New York                  1629507    7.5        
3120 Michigan      Wayne                     1778969   14.9        
3121 Texas         Bexar                     1825502    7.4        
3122 Florida       Broward                   1843152   10.0        
3123 California    Santa Clara               1868149    7.7        
3124 Texas         Tarrant                   1914526    6.9        
3125 Nevada        Clark                     2035572   10.8        
3126 Washington    King                      2045756    6.3        
3127 California    San Bernardino            2094769   12.6        
3128 California    Riverside                 2298032   12.9        
3129 New York      Queens                    2301139    8.6        
3130 Texas         Dallas                    2485003    7.6        
3131 New York      Kings                     2595259   10.0        
3132 Florida       Miami-Dade                2639042   10.0        
3133 California    Orange                    3116069    7.6        
3134 California    San Diego                 3223096    8.7        
3135 Arizona       Maricopa                  4018143    7.7  

### Arrange: descending

In [14]:
counties_selected_scpu %>%
    arrange(desc(population))

state         county                   population unemployment
1    California    Los Angeles              10038388   10.0        
2    Illinois      Cook                      5236393   10.7        
3    Texas         Harris                    4356362    7.5        
4    Arizona       Maricopa                  4018143    7.7        
5    California    San Diego                 3223096    8.7        
6    California    Orange                    3116069    7.6        
7    Florida       Miami-Dade                2639042   10.0        
8    New York      Kings                     2595259   10.0        
9    Texas         Dallas                    2485003    7.6        
10   New York      Queens                    2301139    8.6        
11   California    Riverside                 2298032   12.9        
12   California    San Bernardino            2094769   12.6        
13   Washington    King                      2045756    6.3        
14   Nevada        Clark                     2035572   10.8        
15   Texas         Tarrant                   1914526    6.9        
16   California    Santa Clara               1868149    7.7        
17   Florida       Broward                   1843152   10.0        
18   Texas         Bexar                     1825502    7.4        
19   Michigan      Wayne                     1778969   14.9        
20   New York      New York                  1629507    7.5        
21   California    Alameda                   1584983    8.3        
22   Massachusetts Middlesex                 1556116    6.1        
23   Pennsylvania  Philadelphia              1555072   13.9        
24   New York      Suffolk                   1501373    6.4        
25   California    Sacramento                1465832   11.8        
26   New York      Bronx                     1428357   14.0        
27   Florida       Palm Beach                1378806    9.5        
28   New York      Nassau                    1354612    6.4        
29   Florida       Hillsborough              1302884    9.0        
30   Ohio          Cuyahoga                  1263189   10.6        
...  ...           ...                      ...        ...         
3109 Texas         Terrell                  921         3.7        
3110 Idaho         Clark                    901         3.9        
3111 Colorado      Hinsdale                 874         3.0        
3112 Nebraska      Logan                    851         0.9        
3113 Nebraska      Wheeler                  847         2.9        
3114 Texas         Kent                     821         1.5        
3115 Nebraska      Banner                   820         5.4        
3116 Montana       Treasure                 812         0.8        
3117 South Dakota  Jones                    781         3.6        
3118 Texas         McMullen                 778        14.1        
3119 Utah          Daggett                  776         4.3        
3120 Nebraska      Grant                    769         3.5        
3121 Montana       Golden Valley            756         3.5        
3122 Colorado      Mineral                  733         1.6        
3123 Nebraska      Keya Paha                711         1.1        
3124 Texas         Borden                   705         0.7        
3125 Nebraska      Hooker                   681         2.9        
3126 Nebraska      Thomas                   675         0.0        
3127 North Dakota  Slope                    673         0.0        
3128 Alaska        Yakutat City and Borough 643         7.9        
3129 Colorado      San Juan                 606        13.8        
3130 New Mexico    Harding                  565         6.0        
3131 Texas         Kenedy                   565         0.0        
3132 Nebraska      Blaine                   551         0.7        
3133 Nebraska      Loup                     548         0.7        
3134 Nebraska      Arthur                   448         4.0        
3135 Montana       Petroleum                443         6.6  

### Filter

In [15]:
counties_selected_scpu %>%
    arrange(desc(population)) %>%
    filter(state == 'New York')

state    county       population unemployment
1   New York Kings        2595259    10.0        
2   New York Queens       2301139     8.6        
3   New York New York     1629507     7.5        
4   New York Suffolk      1501373     6.4        
5   New York Bronx        1428357    14.0        
6   New York Nassau       1354612     6.4        
7   New York Westchester   967315     7.6        
8   New York Erie          921584     7.0        
9   New York Monroe        749356     7.7        
10  New York Richmond      472481     6.9        
11  New York Onondaga      468304     7.2        
12  New York Orange        375384     7.3        
13  New York Rockland      320688     7.9        
14  New York Albany        307463     6.4        
15  New York Dutchess      296928     8.7        
16  New York Oneida        233558     7.4        
17  New York Saratoga      223774     6.0        
18  New York Niagara       214150     7.6        
19  New York Broome        198093     7.8        
20  New York Ulster        181300     9.3        
21  New York Rensselaer    159900     7.6        
22  New York Schenectady   154796     7.4        
23  New York Chautauqua    132646     7.8        
24  New York Oswego        121183     9.7        
25  New York Jefferson     118947    10.6        
26  New York St. Lawrence  112011    10.0        
27  New York Ontario       109192     6.4        
28  New York Tompkins      103855     5.9        
29  New York Putnam         99488     6.8        
30  New York Steuben        98665     8.2        
... ...      ...          ...        ...         
33  New York Clinton      81685       6.9        
34  New York Cayuga       79173       7.3        
35  New York Cattaraugus  78962       8.2        
36  New York Sullivan     76330      11.5        
37  New York Madison      72427       5.1        
38  New York Warren       65180       6.6        
39  New York Livingston   64801       5.4        
40  New York Herkimer     64034       8.7        
41  New York Washington   62700       9.5        
42  New York Columbia     62195       6.7        
43  New York Otsego       61399       7.1        
44  New York Genesee      59458       6.7        
45  New York Fulton       54606       9.3        
46  New York Franklin     51280       8.8        
47  New York Tioga        50199       6.9        
48  New York Montgomery   49779       8.9        
49  New York Chenango     49549       7.1        
50  New York Cortland     49043       6.1        
51  New York Greene       48312       9.1        
52  New York Allegany     48070       9.2        
53  New York Delaware     46901       9.5        
54  New York Orleans      42204       9.0        
55  New York Wyoming      41446       7.4        
56  New York Essex        38912       7.9        
57  New York Seneca       35144       5.5        
58  New York Schoharie    31913      10.5        
59  New York Lewis        27124       8.3        
60  New York Yates        25187       6.9        
61  New York Schuyler     18410       6.4        
62  New York Hamilton      4760       9.7

In [16]:
counties_selected_scpu %>%
    arrange(desc(population)) %>%
    filter(unemployment < 6)

state         county              population unemployment
1   Virginia      Fairfax             1128722    4.9         
2   Utah          Salt Lake           1078958    5.8         
3   Hawaii        Honolulu             984178    5.6         
4   Texas         Collin               862215    4.9         
5   Texas         Denton               731851    5.7         
6   Texas         Fort Bend            658331    5.1         
7   Kansas        Johnson              566814    4.5         
8   Maryland      Anne Arundel         555280    5.9         
9   Colorado      Jefferson            552344    5.9         
10  Utah          Utah                 551957    5.5         
11  Nebraska      Douglas              537655    5.7         
12  Wisconsin     Dane                 510198    4.9         
13  Pennsylvania  Chester              509797    5.9         
14  Iowa          Polk                 452369    5.8         
15  Virginia      Prince William       437271    5.4         
16  Minnesota     Dakota               408456    5.0         
17  New Hampshire Hillsborough         403972    5.9         
18  Wisconsin     Waukesha             393873    4.2         
19  Missouri      St. Charles          374805    5.1         
20  Pennsylvania  Westmoreland         361251    5.6         
21  Virginia      Loudoun              351129    4.0         
22  Texas         Brazoria             331741    5.4         
23  Utah          Davis                323374    4.2         
24  Colorado      Douglas              306974    4.1         
25  Maryland      Howard               304115    4.9         
26  New Hampshire Rockingham           299006    5.6         
27  Nebraska      Lancaster            298080    5.1         
28  Indiana       Hamilton             296635    4.3         
29  Maine         Cumberland           286119    5.4         
30  Oklahoma      Cleveland            268614    4.9         
... ...           ...                 ...        ...         
920 Montana       Garfield            1047       0.2         
921 Montana       Wibaux               986       3.0         
922 Alaska        Bristol Bay Borough  970       5.4         
923 North Dakota  Billings             969       1.4         
924 Texas         Roberts              931       1.9         
925 Texas         Terrell              921       3.7         
926 Idaho         Clark                901       3.9         
927 Colorado      Hinsdale             874       3.0         
928 Nebraska      Logan                851       0.9         
929 Nebraska      Wheeler              847       2.9         
930 Texas         Kent                 821       1.5         
931 Nebraska      Banner               820       5.4         
932 Montana       Treasure             812       0.8         
933 South Dakota  Jones                781       3.6         
934 Utah          Daggett              776       4.3         
935 Nebraska      Grant                769       3.5         
936 Montana       Golden Valley        756       3.5         
937 Colorado      Mineral              733       1.6         
938 Nebraska      Keya Paha            711       1.1         
939 Texas         Borden               705       0.7         
940 Nebraska      Hooker               681       2.9         
941 Nebraska      Thomas               675       0.0         
942 North Dakota  Slope                673       0.0         
943 Texas         Kenedy               565       0.0         
944 Nebraska      Blaine               551       0.7         
945 Nebraska      Loup                 548       0.7         
946 Nebraska      Arthur               448       4.0         
947 Nebraska      McPherson            433       0.9         
948 Texas         King                 267       5.1         
949 Hawaii        Kalawao               85       0.0

### Combining conditions

In [17]:
counties_selected_scpu %>%
    arrange(desc(population)) %>%
    filter(state == "New York", unemployment < 6)

state    county     population unemployment
1 New York Tompkins   103855     5.9         
2 New York Chemung     88267     5.4         
3 New York Madison     72427     5.1         
4 New York Livingston  64801     5.4         
5 New York Seneca      35144     5.5

## Exercise

In [18]:
# Arranging observations

counties_selected <- counties %>%
  select(state, county, population, private_work, public_work, self_employed)

counties_selected %>%
  # Add a verb to sort in descending order of public_work
  arrange(desc(public_work))

state        county                     population private_work
1    Hawaii       Kalawao                       85      25.0        
2    Alaska       Yukon-Koyukuk Census Area   5644      33.3        
3    Wisconsin    Menominee                   4451      36.8        
4    North Dakota Sioux                       4380      32.9        
5    South Dakota Todd                        9942      34.4        
6    Alaska       Lake and Peninsula Borough  1474      42.2        
7    California   Lassen                     32645      42.6        
8    South Dakota Buffalo                     2038      48.4        
9    South Dakota Dewey                       5579      34.9        
10   Texas        Kenedy                       565      51.9        
11   North Dakota Rolette                    14498      44.3        
12   Arizona      Apache                     72124      50.2        
13   Alaska       Bethel Census Area         17776      53.2        
14   Utah         Daggett                      776      47.7        
15   Alaska       Nome Census Area            9854      50.6        
16   New Mexico   Mora                        4660      50.7        
17   Alaska       Northwest Arctic Borough    7732      54.0        
18   Montana      Big Horn                   13141      47.7        
19   Nevada       Lincoln                     5194      50.1        
20   South Dakota Jackson                     3237      34.1        
21   New Mexico   Harding                      565      46.3        
22   Montana      Roosevelt                  11072      45.0        
23   California   Sierra                      3021      45.0        
24   Georgia      Calhoun                     6489      53.2        
25   South Dakota Ziebach                     2833      40.9        
26   South Dakota Corson                      4149      36.1        
27   Washington   Ferry                       7652      52.4        
28   Texas        Brewster                    9235      49.2        
29   South Dakota Mellette                    2086      37.3        
30   New Mexico   Catron                      3583      50.3        
...  ...          ...                        ...        ...         
3109 North Dakota Steele                       1979     72.8        
3110 Pennsylvania Chester                    509797     85.9        
3111 Pennsylvania Lehigh                     356756     87.0        
3112 Wisconsin    Ozaukee                     87273     85.5        
3113 Indiana      Dubois                      42291     87.1        
3114 Iowa         Carroll                     20629     81.5        
3115 Missouri     Taney                       53555     82.8        
3116 Illinois     Edwards                      6649     85.8        
3117 Pennsylvania Wyoming                     28147     85.1        
3118 Wisconsin    Waukesha                   393873     87.4        
3119 Michigan     Allegan                    112837     85.2        
3120 Nebraska     Colfax                      10522     83.3        
3121 Pennsylvania Lancaster                  530216     84.7        
3122 Wisconsin    Washington                 132921     87.7        
3123 Pennsylvania Montgomery                 812970     86.2        
3124 Arkansas     Benton                     238198     85.9        
3125 Iowa         Winnebago                   10614     84.3        
3126 Indiana      Elkhart                    200685     87.5        
3127 Indiana      Gibson                      33668     88.0        
3128 Michigan     Kent                       622590     87.3        
3129 Pennsylvania Elk                         31370     87.1        
3130 Indiana      Bartholomew                 79488     87.8        
3131 Illinois     Jersey                      22625     86.6        
3132 Indiana      Clinton                     32835     87.5        
3133 Nebraska     Dakota                      20798     87.3        
3134 Indiana      Kosciusko                   77983     88.3        
313

In [19]:
# Filtering for conditions

counties_selected <- counties %>%
  select(state, county, population)

# Filter for counties with a population above 1000000
counties_selected %>%  
  filter(population > 1000000)

state         county         population
1  Arizona       Maricopa        4018143  
2  California    Alameda         1584983  
3  California    Contra Costa    1096068  
4  California    Los Angeles    10038388  
5  California    Orange          3116069  
6  California    Riverside       2298032  
7  California    Sacramento      1465832  
8  California    San Bernardino  2094769  
9  California    San Diego       3223096  
10 California    Santa Clara     1868149  
11 Florida       Broward         1843152  
12 Florida       Hillsborough    1302884  
13 Florida       Miami-Dade      2639042  
14 Florida       Orange          1229039  
15 Florida       Palm Beach      1378806  
16 Illinois      Cook            5236393  
17 Maryland      Montgomery      1017859  
18 Massachusetts Middlesex       1556116  
19 Michigan      Oakland         1229503  
20 Michigan      Wayne           1778969  
21 Minnesota     Hennepin        1197776  
22 Missouri      St. Louis       1001327  
23 Nevada        Clark           2035572  
24 New York      Bronx           1428357  
25 New York      Kings           2595259  
26 New York      Nassau          1354612  
27 New York      New York        1629507  
28 New York      Queens          2301139  
29 New York      Suffolk         1501373  
30 Ohio          Cuyahoga        1263189  
31 Ohio          Franklin        1215761  
32 Pennsylvania  Allegheny       1231145  
33 Pennsylvania  Philadelphia    1555072  
34 Texas         Bexar           1825502  
35 Texas         Dallas          2485003  
36 Texas         Harris          4356362  
37 Texas         Tarrant         1914526  
38 Texas         Travis          1121645  
39 Utah          Salt Lake       1078958  
40 Virginia      Fairfax         1128722  
41 Washington    King            2045756

In [20]:
# Filter for the counties in the state of California with a population above 1000000
counties_selected %>%
  filter(state == "California", population > 1000000)

state      county         population
1 California Alameda         1584983  
2 California Contra Costa    1096068  
3 California Los Angeles    10038388  
4 California Orange          3116069  
5 California Riverside       2298032  
6 California Sacramento      1465832  
7 California San Bernardino  2094769  
8 California San Diego       3223096  
9 California Santa Clara     1868149

In [21]:
# Filtering and arranging

counties_selected <- counties %>%
  select(state, county, population, private_work, public_work, self_employed)

# Filter for Texas and more than 10000 people; sort in descending order of private_work
counties_selected %>%
  # Filter for Texas and more than 10000 people
  filter(state == "Texas", population > 10000) %>%
  # Sort in descending order of private_work
  arrange(desc(private_work))

state county      population private_work public_work self_employed
1   Texas Gregg        123178    84.7          9.8        5.4          
2   Texas Collin       862215    84.1         10.0        5.8          
3   Texas Dallas      2485003    83.9          9.5        6.4          
4   Texas Harris      4356362    83.4         10.1        6.3          
5   Texas Andrews       16775    83.1          9.6        6.8          
6   Texas Tarrant     1914526    83.1         11.4        5.4          
7   Texas Titus         32553    82.5         10.0        7.4          
8   Texas Denton       731851    82.2         11.9        5.7          
9   Texas Ector        149557    82.0         11.2        6.7          
10  Texas Moore         22281    82.0         11.7        5.9          
11  Texas Jefferson    252872    81.9         13.4        4.5          
12  Texas Fort Bend    658331    81.7         12.5        5.7          
13  Texas Panola        23900    81.5         11.7        6.9          
14  Texas Midland      151290    81.2         11.4        7.1          
15  Texas Potter       122352    81.2         12.4        6.2          
16  Texas Frio          18168    81.1         15.4        3.4          
17  Texas Johnson      155450    80.9         13.3        5.7          
18  Texas Smith        217552    80.9         12.7        6.3          
19  Texas Orange        83217    80.8         13.7        5.4          
20  Texas Harrison      66417    80.6         13.9        5.4          
21  Texas Brazoria     331741    80.5         14.4        5.0          
22  Texas Calhoun       21666    80.5         12.8        6.7          
23  Texas Austin        28886    80.4         11.8        7.7          
24  Texas Montgomery   502586    80.4         11.7        7.8          
25  Texas Jim Wells     41461    80.3         12.6        7.1          
26  Texas Hutchinson    21858    80.2         13.9        5.8          
27  Texas Ochiltree     10642    80.1         10.4        9.2          
28  Texas Victoria      90099    80.1         12.5        7.1          
29  Texas Hardin        55375    80.0         15.7        4.2          
30  Texas Bexar       1825502    79.6         14.8        5.6          
... ...   ...         ...        ...          ...         ...          
140 Texas Robertson    16532     69.8         22.4         7.6         
141 Texas Karnes       14879     69.6         22.0         7.8         
142 Texas Kleberg      32029     69.5         25.9         4.4         
143 Texas Burleson     17293     69.4         21.7         8.4         
144 Texas Leon         16819     69.4         14.5        15.7         
145 Texas Bandera      20796     69.2         19.0        11.6         
146 Texas Bell        326041     69.2         25.7         5.0         
147 Texas San Jacinto  27023     69.1         23.0         7.6         
148 Texas Maverick     56548     68.8         25.0         6.1         
149 Texas Trinity      14405     68.8         17.8        12.1         
150 Texas Jones        19978     68.7         20.5        10.3         
151 Texas Polk         46113     68.7         22.2         8.7         
152 Texas Pecos        15807     68.6         24.5         6.0         
153 Texas Dawson       13542     68.5         21.9         9.5         
154 Texas Milam        24344     68.5         21.3         9.8         
155 Texas Red River    12567     68.5         17.4        13.8         
156 Texas Brazos      205271     68.3         25.6         5.8         
157 Texas Starr        62648     68.2         22.5         8.6         
158 Texas Houston      22949     67.8         22.1         9.8         
159 Texas Reeves       14179     67.8         28.3         3.9         
160 Texas Runnels      10445     67.7         19.9        12.1         
161 Texas Willacy      22002     66.9         23.9         9.1         
162 Texas Uvalde       26952     66.8         23.8         9.4         
163 Texas Wilbarger    13158     66.0       

---
## Mutate
`mutate()` verb adds new variables or change existing variables.

In [24]:
counties_selected_scpu %>%
    mutate(unemployed_population = population * unemployment / 100) %>%
    arrange(desc(unemployed_population))

state        county         population unemployment unemployed_population
1    California   Los Angeles    10038388   10.0         1003838.8            
2    Illinois     Cook            5236393   10.7          560294.1            
3    Texas        Harris          4356362    7.5          326727.2            
4    Arizona      Maricopa        4018143    7.7          309397.0            
5    California   Riverside       2298032   12.9          296446.1            
6    California   San Diego       3223096    8.7          280409.4            
7    Michigan     Wayne           1778969   14.9          265066.4            
8    California   San Bernardino  2094769   12.6          263940.9            
9    Florida      Miami-Dade      2639042   10.0          263904.2            
10   New York     Kings           2595259   10.0          259525.9            
11   California   Orange          3116069    7.6          236821.2            
12   Nevada       Clark           2035572   10.8          219841.8            
13   Pennsylvania Philadelphia    1555072   13.9          216155.0            
14   New York     Bronx           1428357   14.0          199970.0            
15   New York     Queens          2301139    8.6          197898.0            
16   Texas        Dallas          2485003    7.6          188860.2            
17   Florida      Broward         1843152   10.0          184315.2            
18   California   Sacramento      1465832   11.8          172968.2            
19   California   Santa Clara     1868149    7.7          143847.5            
20   Texas        Bexar           1825502    7.4          135087.1            
21   Ohio         Cuyahoga        1263189   10.6          133898.0            
22   Texas        Tarrant         1914526    6.9          132102.3            
23   California   Alameda         1584983    8.3          131553.6            
24   Florida      Palm Beach      1378806    9.5          130986.6            
25   Washington   King            2045756    6.3          128882.6            
26   California   Fresno           956749   13.2          126290.9            
27   New York     New York        1629507    7.5          122213.0            
28   Florida      Hillsborough    1302884    9.0          117259.6            
29   Florida      Orange          1229039    9.1          111842.5            
30   California   Kern             865736   12.7          109948.5            
...  ...          ...            ...        ...          ...                  
3109 Nebraska     Hooker          681       2.9          19.749               
3110 Nebraska     Hayes          1084       1.8          19.512               
3111 South Dakota Harding        1328       1.4          18.592               
3112 South Dakota Jerauld        2031       0.9          18.279               
3113 North Dakota Golden Valley  1822       1.0          18.220               
3114 Nebraska     Arthur          448       4.0          17.920               
3115 Texas        Roberts         931       1.9          17.689               
3116 South Dakota Faulk          2359       0.7          16.513               
3117 Nebraska     Sioux          1249       1.3          16.237               
3118 North Dakota Divide         2314       0.7          16.198               
3119 Montana      McCone         1728       0.8          13.824               
3120 Texas        King            267       5.1          13.617               
3121 North Dakota Billings        969       1.4          13.566               
3122 Texas        Kent            821       1.5          12.315               
3123 South Dakota Douglas        2973       0.4          11.892               
3124 Colorado     Mineral         733       1.6          11.728               
3125 South Dakota Potter         2307       0.5          11.535               
3126 Nebraska     Keya Paha       711       1.1           7.821               
3127 Nebraska     Logan           851       0.9        

## Exercise

In [26]:
# Calculating the number of government employees

counties_selected <- counties %>%
  select(state, county, population, public_work)

counties_selected %>%
  # Add a new column public_workers with the number of people employed in public work
  mutate(public_workers = population * public_work / 100) %>%
  # Sort in descending order of the public_workers column
  arrange(desc(public_workers))

state        county          population public_work public_workers
1    California   Los Angeles     10038388   11.5        1154414.6     
2    Illinois     Cook             5236393   11.5         602185.2     
3    California   San Diego        3223096   14.8         477018.2     
4    Arizona      Maricopa         4018143   11.7         470122.7     
5    Texas        Harris           4356362   10.1         439992.6     
6    New York     Kings            2595259   14.4         373717.3     
7    California   San Bernardino   2094769   16.7         349826.4     
8    California   Riverside        2298032   14.9         342406.8     
9    California   Sacramento       1465832   21.8         319551.4     
10   California   Orange           3116069   10.2         317839.0     
11   New York     Queens           2301139   13.6         312954.9     
12   New York     Suffolk          1501373   18.1         271748.5     
13   Texas        Bexar            1825502   14.8         270174.3     
14   Florida      Miami-Dade       2639042   10.2         269182.3     
15   Washington   King             2045756   12.4         253673.7     
16   Maryland     Prince George's   892816   27.2         242846.0     
17   Virginia     Fairfax          1128722   21.5         242675.2     
18   Texas        Dallas           2485003    9.5         236075.3     
19   New York     Nassau           1354612   16.4         222156.4     
20   Maryland     Montgomery       1017859   21.7         220875.4     
21   California   Alameda          1584983   13.8         218727.7     
22   Texas        Tarrant          1914526   11.4         218256.0     
23   Nevada       Clark            2035572   10.7         217806.2     
24   Hawaii       Honolulu          984178   22.1         217503.3     
25   New York     Bronx            1428357   14.8         211396.8     
26   Florida      Broward          1843152   11.3         208276.2     
27   Pennsylvania Philadelphia     1555072   13.0         202159.4     
28   Arizona      Pima              998537   18.3         182732.3     
29   Ohio         Franklin         1215761   14.8         179932.6     
30   Texas        Travis           1121645   16.0         179463.2     
...  ...          ...             ...        ...         ...           
3109 Nebraska     Rock            1401       14.4        201.744       
3110 Montana      Wheatland       2115        9.5        200.925       
3111 Texas        Glasscock       1180       17.0        200.600       
3112 Texas        McMullen         778       25.6        199.168       
3113 Mississippi  Issaquena       1291       15.4        198.814       
3114 Montana      Carter          1227       16.0        196.320       
3115 Idaho        Clark            901       21.5        193.715       
3116 Montana      Golden Valley    756       23.2        175.392       
3117 North Dakota Sheridan        1337       13.0        173.810       
3118 North Dakota Billings         969       17.6        170.544       
3119 North Dakota Steele          1979        8.5        168.215       
3120 Nebraska     Hooker           681       23.6        160.716       
3121 Nebraska     Banner           820       19.1        156.620       
3122 Nebraska     Thomas           675       22.3        150.525       
3123 Nebraska     Wheeler          847       16.8        142.296       
3124 Colorado     Mineral          733       18.8        137.804       
3125 Nebraska     Logan            851       14.0        119.140       
3126 Nebraska     Keya Paha        711       16.6        118.026       
3127 South Dakota Jones            781       14.7        114.807       
3128 Nebraska     Loup             548       19.5        106.860       
3129 North Dakota Slope            673       15.5        104.315       
3130 Montana      Treasure         812       12.4        100.688       
3131 Montana      Petroleum        443       22.0         97.460       
3132 Nebraska     Blaine           551       

In [27]:
# Calculating the percentage of women in a county

counties_selected <- counties %>%
  # Select the columns state, county, population, men, and women
  select(state, county, population, men, women)

counties_selected %>%
  # Calculate proportion_women as the fraction of the population made up of women
  mutate(proportion_women = women / population)

state     county      population men    women  proportion_women
1    Alabama   Autauga      55221     26745  28476  0.5156734       
2    Alabama   Baldwin     195121     95314  99807  0.5115134       
3    Alabama   Barbour      26932     14497  12435  0.4617184       
4    Alabama   Bibb         22604     12073  10531  0.4658910       
5    Alabama   Blount       57710     28512  29198  0.5059435       
6    Alabama   Bullock      10678      5660   5018  0.4699382       
7    Alabama   Butler       20354      9502  10852  0.5331630       
8    Alabama   Calhoun     116648     56274  60374  0.5175742       
9    Alabama   Chambers     34079     16258  17821  0.5229320       
10   Alabama   Cherokee     26008     12975  13033  0.5011150       
11   Alabama   Chilton      43819     21619  22200  0.5066295       
12   Alabama   Choctaw      13395      6382   7013  0.5235536       
13   Alabama   Clarke       25070     11834  13236  0.5279617       
14   Alabama   Clay         13537      6671   6866  0.5072025       
15   Alabama   Cleburne     15002      7334   7668  0.5111318       
16   Alabama   Coffee       50884     25174  25710  0.5052669       
17   Alabama   Colbert      54444     26303  28141  0.5168797       
18   Alabama   Conecuh      12865      6176   6689  0.5199378       
19   Alabama   Coosa        11027      5579   5448  0.4940600       
20   Alabama   Covington    37886     18339  19547  0.5159426       
21   Alabama   Crenshaw     13938      6863   7075  0.5076051       
22   Alabama   Cullman      80965     40081  40884  0.5049589       
23   Alabama   Dale         49866     24708  25158  0.5045121       
24   Alabama   Dallas       42154     19450  22704  0.5385966       
25   Alabama   DeKalb       71068     35474  35594  0.5008443       
26   Alabama   Elmore       80763     39362  41401  0.5126234       
27   Alabama   Escambia     37935     19524  18411  0.4853302       
28   Alabama   Etowah      103766     50207  53559  0.5161517       
29   Alabama   Fayette      16896      8477   8419  0.4982836       
30   Alabama   Franklin     31634     15311  16323  0.5159954       
...  ...       ...         ...        ...    ...    ...             
3109 Wisconsin Washburn     15700       7799   7901 0.5032484       
3110 Wisconsin Washington  132921      65880  67041 0.5043673       
3111 Wisconsin Waukesha    393873     193441 200432 0.5088747       
3112 Wisconsin Waupaca      52125      26378  25747 0.4939472       
3113 Wisconsin Waushara     24321      12783  11538 0.4744048       
3114 Wisconsin Winnebago   169004      85019  83985 0.4969409       
3115 Wisconsin Wood         74012      36594  37418 0.5055667       
3116 Wyoming   Albany       37565      19692  17873 0.4757886       
3117 Wyoming   Big Horn     11895       6002   5893 0.4954182       
3118 Wyoming   Campbell     48013      24914  23099 0.4810989       
3119 Wyoming   Carbon       15739       8592   7147 0.4540949       
3120 Wyoming   Converse     14101       7118   6983 0.4952131       
3121 Wyoming   Crook         7229       3716   3513 0.4859593       
3122 Wyoming   Fremont      40755      20358  20397 0.5004785       
3123 Wyoming   Goshen       13544       7180   6364 0.4698760       
3124 Wyoming   Hot Springs   4809       2394   2415 0.5021834       
3125 Wyoming   Johnson       8600       4525   4075 0.4738372       
3126 Wyoming   Laramie      95431      47911  47520 0.4979514       
3127 Wyoming   Lincoln      18316       9356   8960 0.4891898       
3128 Wyoming   Natrona      80011      40495  39516 0.4938821       
3129 Wyoming   Niobrara      2516       1164   1352 0.5373609       
3130 Wyoming   Park         28985      14230  14755 0.5090564       
3131 Wyoming   Platte        8750       4495   4255 0.4862857       
3132 Wyoming   Sheridan     29738      14695  15043 0.5058511       
3133 Wyoming   Sublette     10117       5353   4764 0.4708906       
3134 Wyoming   Sweetwater   44772      23359  21413 0.4782677       
313

In [28]:
# Select, mutate, filter, and arrange

counties %>%
  select(state, county, population, men, women) %>%
  # Add the proportion_men variable
  mutate(proportion_men = men / population) %>%
  # Filter for population of at least 10,000
  filter(population >= 10000) %>%
  # Arrange proportion of men in descending order 
  arrange(desc(proportion_men))

state          county                population men   women proportion_men
1    Virginia       Sussex                11864       8130  3734 0.6852664     
2    California     Lassen                32645      21818 10827 0.6683412     
3    Georgia        Chattahoochee         11914       7940  3974 0.6664428     
4    Louisiana      West Feliciana        15415      10228  5187 0.6635096     
5    Florida        Union                 15191       9830  5361 0.6470937     
6    Texas          Jones                 19978      12652  7326 0.6332966     
7    Missouri       DeKalb                12782       8080  4702 0.6321389     
8    Texas          Madison               13838       8648  5190 0.6249458     
9    Virginia       Greensville           11760       7303  4457 0.6210034     
10   Texas          Anderson              57915      35469 22446 0.6124320     
11   Arkansas       Lincoln               14062       8596  5466 0.6112928     
12   Texas          Bee                   32659      19722 12937 0.6038764     
13   Florida        Hamilton              14395       8671  5724 0.6023619     
14   Illinois       Lawrence              16665       9997  6668 0.5998800     
15   Mississippi    Tallahatchie          14959       8907  6052 0.5954275     
16   Mississippi    Greene                14129       8409  5720 0.5951589     
17   Texas          Karnes                14879       8799  6080 0.5913704     
18   Texas          Frio                  18168      10723  7445 0.5902136     
19   Florida        Gulf                  15785       9235  6550 0.5850491     
20   Florida        Franklin              11628       6800  4828 0.5847953     
21   Texas          Walker                69330      40484 28846 0.5839319     
22   Georgia        Telfair               16416       9502  6914 0.5788255     
23   Colorado       Fremont               46809      27003 19806 0.5768762     
24   Louisiana      Allen                 25653      14784 10869 0.5763069     
25   Ohio           Noble                 14508       8359  6149 0.5761649     
26   Georgia        Tattnall              25302      14529 10773 0.5742234     
27   Louisiana      Claiborne             16639       9524  7115 0.5723902     
28   Texas          Pecos                 15807       9003  6804 0.5695578     
29   Missouri       Pulaski               53443      30390 23053 0.5686432     
30   Texas          Howard                36105      20496 15609 0.5676776     
...  ...            ...                   ...        ...   ...   ...           
2408 South Carolina Greenwood             69771      32567 37204 0.4667699     
2409 Virginia       Fredericksburg city   27395      12783 14612 0.4666180     
2410 South Carolina Barnwell              22098      10290 11808 0.4656530     
2411 Virginia       Bristol city          17524       8158  9366 0.4655330     
2412 Alabama        Marengo               20306       9452 10854 0.4654782     
2413 Georgia        Dougherty             93310      43432 49878 0.4654592     
2414 Mississippi    Clay                  20279       9435 10844 0.4652596     
2415 Virginia       Colonial Heights city 17515       8142  9373 0.4648587     
2416 North Carolina Edgecombe             55280      25693 29587 0.4647793     
2417 Georgia        Evans                 10814       5024  5790 0.4645829     
2418 Arkansas       Desha                 12379       5741  6638 0.4637693     
2419 Alabama        Perry                 10038       4651  5387 0.4633393     
2420 Georgia        Early                 10579       4899  5680 0.4630872     
2421 Virginia       Essex                 11151       5158  5993 0.4625594     
2422 Mississippi    Coahoma               25254      11681 13573 0.4625406     
2423 Alabama        Dallas                42154      19450 22704 0.4614034     
2424 North Carolina Washington            12668       5828  6840 0.4600568     
2425 Virginia       Petersburg city       32123      14773 17350 0.4598886     
2426 

*Notice Sussex County in Virginia is more than two thirds male: this is because of two men's prisons in the county.*

---
---
# 2. Aggregating Data
Now that you know how to transform your data, you'll want to know more about how to aggregate your data to make it more interpretable. You'll learn a number of functions you can use to take many observations in your data and summarize them, including count, group_by, summarize, ungroup, and top_n.

## The count verb

In [29]:
counties %>%
    count()

n   
1 3138

### Count variable

In [30]:
counties %>%
    count(state)

state          n  
1  Alabama         67
2  Alaska          28
3  Arizona         15
4  Arkansas        75
5  California      58
6  Colorado        64
7  Connecticut      8
8  Delaware         3
9  Florida         67
10 Georgia        159
11 Hawaii           5
12 Idaho           44
13 Illinois       102
14 Indiana         92
15 Iowa            99
16 Kansas         105
17 Kentucky       120
18 Louisiana       64
19 Maine           16
20 Maryland        24
21 Massachusetts   14
22 Michigan        83
23 Minnesota       87
24 Mississippi     82
25 Missouri       115
26 Montana         56
27 Nebraska        93
28 Nevada          17
29 New Hampshire   10
30 New Jersey      21
31 New Mexico      33
32 New York        62
33 North Carolina 100
34 North Dakota    53
35 Ohio            88
36 Oklahoma        77
37 Oregon          36
38 Pennsylvania    67
39 Rhode Island     5
40 South Carolina  46
41 South Dakota    65
42 Tennessee       95
43 Texas          253
44 Utah            29
45 Vermont         14
46 Virginia       133
47 Washington      39
48 West Virginia   55
49 Wisconsin       72
50 Wyoming         23

### Count and sort

In [31]:
counties %>%
    count(state, sort = TRUE)

state          n  
1  Texas          253
2  Georgia        159
3  Virginia       133
4  Kentucky       120
5  Missouri       115
6  Kansas         105
7  Illinois       102
8  North Carolina 100
9  Iowa            99
10 Tennessee       95
11 Nebraska        93
12 Indiana         92
13 Ohio            88
14 Minnesota       87
15 Michigan        83
16 Mississippi     82
17 Oklahoma        77
18 Arkansas        75
19 Wisconsin       72
20 Alabama         67
21 Florida         67
22 Pennsylvania    67
23 South Dakota    65
24 Colorado        64
25 Louisiana       64
26 New York        62
27 California      58
28 Montana         56
29 West Virginia   55
30 North Dakota    53
31 South Carolina  46
32 Idaho           44
33 Washington      39
34 Oregon          36
35 New Mexico      33
36 Utah            29
37 Alaska          28
38 Maryland        24
39 Wyoming         23
40 New Jersey      21
41 Nevada          17
42 Maine           16
43 Arizona         15
44 Massachusetts   14
45 Vermont         14
46 New Hampshire   10
47 Connecticut      8
48 Hawaii           5
49 Rhode Island     5
50 Delaware         3

### Count population/Add weight
Add the argument `wt`, which stands for 'weight'.

In [32]:
counties %>%
    count(state, wt = population, sort = TRUE)

state          n       
1  California     38421464
2  Texas          26538497
3  New York       19673174
4  Florida        19645772
5  Illinois       12873761
6  Pennsylvania   12779559
7  Ohio           11575977
8  Georgia        10006693
9  Michigan        9900571
10 North Carolina  9845333
11 New Jersey      8904413
12 Virginia        8256630
13 Washington      6985464
14 Massachusetts   6705586
15 Arizona         6641928
16 Indiana         6568645
17 Tennessee       6499615
18 Missouri        6045448
19 Maryland        5930538
20 Wisconsin       5742117
21 Minnesota       5419171
22 Colorado        5278906
23 Alabama         4830620
24 South Carolina  4777576
25 Louisiana       4625253
26 Kentucky        4397353
27 Oregon          3939233
28 Oklahoma        3849733
29 Connecticut     3593222
30 Iowa            3093526
31 Mississippi     2988081
32 Arkansas        2958208
33 Utah            2903379
34 Kansas          2892987
35 Nevada          2798636
36 New Mexico      2084117
37 Nebraska        1869365
38 West Virginia   1851420
39 Idaho           1616547
40 Hawaii          1406299
41 Maine           1329100
42 New Hampshire   1324201
43 Rhode Island    1053661
44 Montana         1014699
45 Delaware         926454
46 South Dakota     829037
47 Alaska           725461
48 North Dakota     721640
49 Vermont          626604
50 Wyoming          579679

## Exercise

In [33]:
# Counting by region

counties_selected <- counties %>%
  select(county, region, state, population, citizens)

# Use count to find the number of counties in each region
counties_selected %>%
  count(region, sort = TRUE)

region        n   
1 South         1420
2 North Central 1054
3 West           447
4 Northeast      217

In [34]:
# Counting citizens by state

# Find number of counties per state, weighted by citizens, sorted in descending order
counties_selected %>%
  count(state, wt = citizens, sort = TRUE)

state          n       
1  California     24280349
2  Texas          16864864
3  Florida        13933052
4  New York       13531404
5  Pennsylvania    9710416
6  Illinois        8979999
7  Ohio            8709050
8  Michigan        7380136
9  North Carolina  7107998
10 Georgia         6978660
11 New Jersey      6053893
12 Virginia        5948625
13 Washington      4937212
14 Massachusetts   4850598
15 Tennessee       4828366
16 Indiana         4801113
17 Arizona         4526594
18 Missouri        4525035
19 Wisconsin       4294321
20 Maryland        4182241
21 Minnesota       3950807
22 Colorado        3750953
23 Alabama         3620994
24 South Carolina  3566508
25 Louisiana       3410634
26 Kentucky        3297108
27 Oregon          2867670
28 Oklahoma        2768561
29 Connecticut     2574178
30 Iowa            2285126
31 Mississippi     2210424
32 Arkansas        2164083
33 Kansas          2053919
34 Utah            1868008
35 Nevada          1863799
36 New Mexico      1457632
37 West Virginia   1455848
38 Nebraska        1333860
39 Idaho           1130550
40 Maine           1048274
41 New Hampshire   1020130
42 Hawaii          1001729
43 Montana          781250
44 Rhode Island     776565
45 Delaware         681606
46 South Dakota     612867
47 North Dakota     546486
48 Alaska           519266
49 Vermont          493124
50 Wyoming          430026

*California is the state with the most citizens.*

In [36]:
# Mutating and counting
counties_selected <- counties %>%
  select(county, region, state, population, walk)

counties_selected %>%
  # Add population_walk containing the total number of people who walk to work 
  mutate(population_walk = population * walk / 100) %>%
  # Count weighted by the new column, sort in descending order
  count(state, wt = population_walk, sort = TRUE)

state          n         
1  New York       1237938.17
2  California     1017963.68
3  Pennsylvania    505397.19
4  Texas           430783.43
5  Illinois        400345.60
6  Massachusetts   316765.03
7  Florida         284722.87
8  New Jersey      273047.19
9  Ohio            266910.98
10 Washington      239764.32
11 Michigan        218555.26
12 Wisconsin       190405.65
13 Virginia        189785.38
14 North Carolina  174886.80
15 Georgia         158961.85
16 Oregon          158401.83
17 Colorado        156067.67
18 Minnesota       154275.49
19 Maryland        146489.35
20 Arizona         139854.32
21 Indiana         137587.55
22 Missouri        119358.22
23 Connecticut     108348.18
24 Iowa            108177.69
25 South Carolina  103450.22
26 Kentucky        101761.91
27 Louisiana        87785.96
28 Tennessee        85869.12
29 Utah             74593.11
30 Oklahoma         68707.91
31 Kansas           68252.54
32 Hawaii           63123.57
33 Nevada           61084.79
34 Alaska           56672.53
35 Alabama          54167.37
36 Nebraska         53082.82
37 Maine            52921.92
38 Arkansas         52499.64
39 West Virginia    51493.53
40 Montana          49894.82
41 Idaho            48935.74
42 New Mexico       47030.15
43 Mississippi      45874.36
44 Rhode Island     39368.18
45 New Hampshire    38290.99
46 Vermont          35655.86
47 South Dakota     33351.02
48 North Dakota     28780.16
49 Wyoming          21981.33
50 Delaware         20084.09

*While California had the largest total population, New York state has the largest number of people who walk to work.*

---
## The group by, summarize and ungroup verbs
`count` is a special case of a more general set of verbs: group by and summarize.

### Summarize
The summarize verb takes many observations and turns them into one observation.

In [37]:
counties %>%
    summarize(total_population = sum(population))

total_population
1 315845353

### Aggregate and summarize

In [38]:
counties %>%
    summarize(total_population = sum(population), 
              average_unemployment = mean(unemployment))

total_population average_unemployment
1 315845353        7.798343

### Summary functions
- `sum()`
- `mean()`
- `median()`
- `min()`
- `max()`
- `n()` : the size of the group

###  Aggregate within groups + Arrange

In [40]:
counties %>%
    group_by(state) %>%
    summarize(total_pop = sum(population),
              average_unemployment = mean(unemployment)) %>%
    arrange(desc(average_unemployment))

state          total_pop average_unemployment
1  Mississippi     2988081  12.024390           
2  Arizona         6641928  11.966667           
3  South Carolina  4777576  11.330435           
4  Alabama         4830620  11.310448           
5  California     38421464  10.800000           
6  Nevada          2798636  10.476471           
7  North Carolina  9845333  10.452000           
8  Florida        19645772  10.383582           
9  Georgia        10006693   9.974214           
10 Michigan        9900571   9.962651           
11 Oregon          3939233   9.927778           
12 Tennessee       6499615   9.503158           
13 Kentucky        4397353   9.337500           
14 Louisiana       4625253   9.232812           
15 Alaska           725461   9.185714           
16 Arkansas        2958208   8.982667           
17 New Jersey      8904413   8.966667           
18 Washington      6985464   8.841026           
19 West Virginia   1851420   8.670909           
20 New Mexico      2084117   8.469697           
21 Connecticut     3593222   8.162500           
22 Illinois       12873761   7.974510           
23 Ohio           11575977   7.957955           
24 Delaware         926454   7.933333           
25 New York       19673174   7.858065           
26 Missouri        6045448   7.605217           
27 Rhode Island    1053661   7.560000           
28 Maryland        5930538   7.475000           
29 Colorado        5278906   7.456250           
30 Maine           1329100   7.437500           
31 Indiana         6568645   7.333696           
32 Massachusetts   6705586   7.292857           
33 Virginia        8256630   7.287970           
34 Pennsylvania   12779559   7.229851           
35 Idaho           1616547   7.081818           
36 Texas          26538497   6.705138           
37 Oklahoma        3849733   6.329870           
38 Wisconsin       5742117   6.172222           
39 Utah            2903379   5.958621           
40 Vermont          626604   5.935714           
41 New Hampshire   1324201   5.810000           
42 Montana         1014699   5.553571           
43 South Dakota     829037   5.249231           
44 Hawaii          1406299   5.200000           
45 Minnesota       5419171   4.975862           
46 Kansas          2892987   4.648571           
47 Iowa            3093526   4.608081           
48 Wyoming          579679   4.591304           
49 Nebraska        1869365   3.345161           
50 North Dakota     721640   2.681132

### Metro column & Group by

In [42]:
counties %>%
    group_by(state, metro) %>%
    summarize(total_pop = sum(population))

`summarise()` has grouped output by 'state'. You can override using the `.groups` argument.


state          metro    total_pop
1   Alabama        Metro     3671377 
2   Alabama        Nonmetro  1159243 
3   Alaska         Metro      494990 
4   Alaska         Nonmetro   230471 
5   Arizona        Metro     6295145 
6   Arizona        Nonmetro   346783 
7   Arkansas       Metro     1806867 
8   Arkansas       Nonmetro  1151341 
9   California     Metro    37587429 
10  California     Nonmetro   834035 
11  Colorado       Metro     4590896 
12  Colorado       Nonmetro   688010 
13  Connecticut    Metro     3406918 
14  Connecticut    Nonmetro   186304 
15  Delaware       Metro      926454 
16  Florida        Metro    18941821 
17  Florida        Nonmetro   703951 
18  Georgia        Metro     8233886 
19  Georgia        Nonmetro  1772807 
20  Hawaii         Metro     1145126 
21  Hawaii         Nonmetro   261173 
22  Idaho          Metro     1069978 
23  Idaho          Nonmetro   546569 
24  Illinois       Metro    11368647 
25  Illinois       Nonmetro  1505114 
26  Indiana        Metro     5095864 
27  Indiana        Nonmetro  1472781 
28  Iowa           Metro     1804496 
29  Iowa           Nonmetro  1289030 
30  Kansas         Metro     1941587 
... ...            ...      ...      
68  Ohio           Nonmetro  2377054 
69  Oklahoma       Metro     2504090 
70  Oklahoma       Nonmetro  1345643 
71  Oregon         Metro     3286017 
72  Oregon         Nonmetro   653216 
73  Pennsylvania   Metro    11285024 
74  Pennsylvania   Nonmetro  1494535 
75  Rhode Island   Metro     1053661 
76  South Carolina Metro     4023755 
77  South Carolina Nonmetro   753821 
78  South Dakota   Metro      398433 
79  South Dakota   Nonmetro   430604 
80  Tennessee      Metro     5007264 
81  Tennessee      Nonmetro  1492351 
82  Texas          Metro    23510680 
83  Texas          Nonmetro  3027817 
84  Utah           Metro     2591224 
85  Utah           Nonmetro   312155 
86  Vermont        Metro      215081 
87  Vermont        Nonmetro   411523 
88  Virginia       Metro     7195353 
89  Virginia       Nonmetro  1061277 
90  Washington     Metro     6272805 
91  Washington     Nonmetro   712659 
92  West Virginia  Metro     1137523 
93  West Virginia  Nonmetro   713897 
94  Wisconsin      Metro     4237349 
95  Wisconsin      Nonmetro  1504768 
96  Wyoming        Metro      175442 
97  Wyoming        Nonmetro   404237

### Ungroup
If not want to keep state as a group, add another dplyr verb: `ungroup()`.

In [43]:
counties %>%
    group_by(state, metro) %>%
    summarize(total_pop = sum(population)) %>%
    ungroup()

`summarise()` has grouped output by 'state'. You can override using the `.groups` argument.


state          metro    total_pop
1   Alabama        Metro     3671377 
2   Alabama        Nonmetro  1159243 
3   Alaska         Metro      494990 
4   Alaska         Nonmetro   230471 
5   Arizona        Metro     6295145 
6   Arizona        Nonmetro   346783 
7   Arkansas       Metro     1806867 
8   Arkansas       Nonmetro  1151341 
9   California     Metro    37587429 
10  California     Nonmetro   834035 
11  Colorado       Metro     4590896 
12  Colorado       Nonmetro   688010 
13  Connecticut    Metro     3406918 
14  Connecticut    Nonmetro   186304 
15  Delaware       Metro      926454 
16  Florida        Metro    18941821 
17  Florida        Nonmetro   703951 
18  Georgia        Metro     8233886 
19  Georgia        Nonmetro  1772807 
20  Hawaii         Metro     1145126 
21  Hawaii         Nonmetro   261173 
22  Idaho          Metro     1069978 
23  Idaho          Nonmetro   546569 
24  Illinois       Metro    11368647 
25  Illinois       Nonmetro  1505114 
26  Indiana        Metro     5095864 
27  Indiana        Nonmetro  1472781 
28  Iowa           Metro     1804496 
29  Iowa           Nonmetro  1289030 
30  Kansas         Metro     1941587 
... ...            ...      ...      
68  Ohio           Nonmetro  2377054 
69  Oklahoma       Metro     2504090 
70  Oklahoma       Nonmetro  1345643 
71  Oregon         Metro     3286017 
72  Oregon         Nonmetro   653216 
73  Pennsylvania   Metro    11285024 
74  Pennsylvania   Nonmetro  1494535 
75  Rhode Island   Metro     1053661 
76  South Carolina Metro     4023755 
77  South Carolina Nonmetro   753821 
78  South Dakota   Metro      398433 
79  South Dakota   Nonmetro   430604 
80  Tennessee      Metro     5007264 
81  Tennessee      Nonmetro  1492351 
82  Texas          Metro    23510680 
83  Texas          Nonmetro  3027817 
84  Utah           Metro     2591224 
85  Utah           Nonmetro   312155 
86  Vermont        Metro      215081 
87  Vermont        Nonmetro   411523 
88  Virginia       Metro     7195353 
89  Virginia       Nonmetro  1061277 
90  Washington     Metro     6272805 
91  Washington     Nonmetro   712659 
92  West Virginia  Metro     1137523 
93  West Virginia  Nonmetro   713897 
94  Wisconsin      Metro     4237349 
95  Wisconsin      Nonmetro  1504768 
96  Wyoming        Metro      175442 
97  Wyoming        Nonmetro   404237

## Exercise

In [44]:
# Summarizing

counties_selected <- counties %>%
  select(county, population, income, unemployment)

# Summarize to find minimum population, maximum unemployment, and average income
counties_selected %>%
  summarize(min_population = min(population), 
            max_unemployment = max(unemployment), 
            average_income = mean(income))

min_population max_unemployment average_income
1 85             29.4             46832

In [45]:
# Summarizing by state

counties_selected <- counties %>%
  select(state, county, population, land_area)

counties_selected %>%
  # Group by state 
  group_by(state) %>%
  # Find the total area and population
  summarise(total_area = sum(land_area), 
            total_population = sum(population)) %>%
  # Add a density column
  mutate(density = total_population / total_area) %>%
  # Sort by density in descending order
  arrange(desc(density))  

state          total_area total_population density    
1  New Jersey       7354.22   8904413         1210.789587
2  Rhode Island     1033.82   1053661         1019.191929
3  Massachusetts    7800.08   6705586          859.681696
4  Connecticut      4842.36   3593222          742.039419
5  Maryland         9707.23   5930538          610.940299
6  Delaware         1948.55    926454          475.458161
7  New York        47126.43  19673174          417.455216
8  Florida         53624.78  19645772          366.356226
9  Pennsylvania    44742.71  12779559          285.623267
10 Ohio            40860.73  11575977          283.303235
11 California     155779.21  38421464          246.640511
12 Illinois        55519.02  12873761          231.880192
13 Hawaii           6422.63   1406299          218.959990
14 Virginia        39483.32   8256630          209.116913
15 North Carolina  48618.01   9845333          202.503825
16 Indiana         35826.16   6568645          183.347727
17 Michigan        56538.92   9900571          175.110720
18 Georgia         57513.54  10006693          173.988473
19 South Carolina  30060.72   4777576          158.930857
20 Tennessee       41235.00   6499615          157.623742
21 New Hampshire    8952.66   1324201          147.911459
22 Kentucky        39486.38   4397353          111.363792
23 Louisiana       43203.93   4625253          107.056303
24 Wisconsin       54157.87   5742117          106.025532
25 Washington      66455.55   6985464          105.114832
26 Texas          260562.87  26538497          101.850647
27 Alabama         50645.39   4830620           95.381238
28 Missouri        68741.54   6045448           87.944611
29 West Virginia   24038.21   1851420           77.019878
30 Minnesota       79626.79   5419171           68.057133
31 Vermont          9216.67    626604           67.985943
32 Mississippi     46923.29   2988081           63.680126
33 Arizona        113594.09   6641928           58.470718
34 Arkansas        52035.48   2958208           56.849826
35 Oklahoma        68594.99   3849733           56.122656
36 Iowa            55857.19   3093526           55.382772
37 Colorado       103641.93   5278906           50.934077
38 Maine           30842.92   1329100           43.092548
39 Oregon          95988.05   3939233           41.038786
40 Kansas          81758.76   2892987           35.384429
41 Utah            82169.62   2903379           35.333971
42 Nevada         109781.18   2798636           25.492858
43 Nebraska        76824.22   1869365           24.333016
44 Idaho           82643.19   1616547           19.560559
45 New Mexico     121298.17   2084117           17.181768
46 South Dakota    73717.14    829037           11.246191
47 North Dakota    69000.81    721640           10.458428
48 Montana        145545.84   1014699            6.971680
49 Wyoming         97093.14    579679            5.970339
50 Alaska         553559.51    725461            1.310538

*Looks like New Jersey and Rhode Island are the “most crowded” of the US states, with more than a thousand people per square mile.*

In [46]:
# Summarizing by state and region

counties_selected <- counties %>%
  select(region, state, county, population)

counties_selected %>%
  # Group and summarize to find the total population
  group_by(region, state) %>%
  summarize(total_pop = sum(population)) %>%
  # Calculate the average_pop and median_pop columns 
  summarize(average_pop = mean(total_pop), median_pop = median(total_pop))

`summarise()` has grouped output by 'region'. You can override using the `.groups` argument.


region        average_pop median_pop
1 North Central 5627687     5580644   
2 Northeast     6221058     3593222   
3 South         7370486     4804098   
4 West          5722755     2798636

*It looks like the South has the highest* `average_pop` *of 7370486, while the North Central region has the highest* `median_pop` *of 5580644.*

---
## The top_n verb
dplyr's `top n` is very useful for keeping the most extreme observations from each group.

### top_n
Like `summarize()`, `top_n()` operates on a grouped table. The function takes two arguments: the number of observations you want from each group, and the column you want to weight by.

For example, `group_by(state)` and then `top_n(1, population)` would find the county with the highest population in each state.

In [47]:
counties_selected <- counties %>%
    select(state, county, population, unemployment, income)

counties_selected %>%
    group_by(state) %>%
    top_n(1, population)

state          county                 population unemployment income
1  Alabama        Jefferson                659026    9.1          45610
2  Alaska         Anchorage Municipality   299107    6.7          78326
3  Arizona        Maricopa                4018143    7.7          54229
4  Arkansas       Pulaski                  390463    7.5          46140
5  California     Los Angeles            10038388   10.0          56196
6  Colorado       El Paso                  655024    8.4          58206
7  Connecticut    Fairfield                939983    9.0          84233
8  Delaware       New Castle               549643    7.4          65476
9  Florida        Miami-Dade              2639042   10.0          43129
10 Georgia        Fulton                   983903    9.9          57207
11 Hawaii         Honolulu                 984178    5.6          74460
12 Idaho          Ada                      417501    6.6          56356
13 Illinois       Cook                    5236393   10.7          55251
14 Indiana        Marion                   926335   10.1          42168
15 Iowa           Polk                     452369    5.8          60061
16 Kansas         Johnson                  566814    4.5          76113
17 Kentucky       Jefferson                755809    8.4          48695
18 Louisiana      East Baton Rouge         444690    7.6          49285
19 Maine          Cumberland               286119    5.4          60051
20 Maryland       Montgomery              1017859    6.1          99435
21 Massachusetts  Middlesex               1556116    6.1          85118
22 Michigan       Wayne                   1778969   14.9          41210
23 Minnesota      Hennepin                1197776    6.0          65834
24 Mississippi    Hinds                    245874   10.3          37324
25 Missouri       St. Louis               1001327    7.4          59755
26 Montana        Yellowstone              153692    4.1          52802
27 Nebraska       Douglas                  537655    5.7          54659
28 Nevada         Clark                   2035572   10.8          51575
29 New Hampshire  Hillsborough             403972    5.9          71244
30 New Jersey     Bergen                   926330    6.5          85806
31 New Mexico     Bernalillo               673943    8.4          47725
32 New York       Kings                   2595259   10.0          48201
33 North Carolina Mecklenburg              990288    8.9          56854
34 North Dakota   Cass                     162500    3.2          53289
35 Ohio           Cuyahoga                1263189   10.6          44190
36 Oklahoma       Oklahoma                 754480    6.0          47437
37 Oregon         Multnomah                768418    8.8          54102
38 Pennsylvania   Philadelphia            1555072   13.9          38253
39 Rhode Island   Providence               630459    9.5          49743
40 South Carolina Greenville               474903    7.6          50540
41 South Dakota   Minnehaha                178942    4.4          53525
42 Tennessee      Shelby                   937750   10.3          46224
43 Texas          Harris                  4356362    7.5          54457
44 Utah           Salt Lake               1078958    5.8          62117
45 Vermont        Chittenden               159711    5.0          65350
46 Virginia       Fairfax                 1128722    4.9         112552
47 Washington     King                    2045756    6.3          75302
48 West Virginia  Kanawha                  190781    7.0          45882
49 Wisconsin      Milwaukee                955939    9.3          43873
50 Wyoming        Laramie                   95431    5.0          60706

Jefferson is the highest population county in Alabama with a population of 659 thousand. Notice that it kept other columns in this table, in this case, unemployment and income.

### Highest unemployment

In [48]:
counties_selected %>%
    group_by(state) %>%
    top_n(1, unemployment)

state          county                   population unemployment income
1  Alabama        Conecuh                    12865    22.6         24900 
2  Alaska         Northwest Arctic Borough    7732    21.9         63648 
3  Arizona        Navajo                    107656    19.8         35921 
4  Arkansas       Phillips                   20391    18.1         26844 
5  California     Imperial                  178206    17.4         41079 
6  Colorado       Crowley                     5551    27.0         31151 
7  Connecticut    New Haven                 862224     9.5         61640 
8  Delaware       Kent                      169509     8.4         54976 
9  Florida        Hamilton                   14395    15.8         35048 
10 Georgia        Taylor                      8401    20.6         28143 
11 Hawaii         Hawaii                    191482     8.1         52108 
12 Idaho          Gem                        16731    14.2         40828 
13 Illinois       Hardin                      4228    18.1         38838 
14 Indiana        Fayette                    23773    12.1         39379 
15 Iowa           Montgomery                 10465     9.6         42418 
16 Kansas         Wyandotte                 160806    11.2         40113 
17 Kentucky       Martin                     12631    20.6         25795 
18 Louisiana      East Carroll                7501    19.2         21099 
19 Maine          Piscataquis                17156    10.5         37495 
20 Maryland       Baltimore city            622454    13.1         42241 
21 Massachusetts  Hampden                   468041    10.0         50461 
22 Michigan       Montmorency                 9401    16.0         36250 
23 Minnesota      Beltrami                   45434    10.6         44757 
24 Mississippi    Humphreys                   8984    26.8         23216 
25 Missouri       Dallas                     16564    13.2         38062 
26 Montana        Big Horn                   13141    19.9         41622 
27 Nebraska       Thurston                    6946    14.4         41266 
28 Nevada         Mineral                     4566    15.4         38923 
29 New Hampshire  Coos                       31870     6.9         42312 
30 New Jersey     Essex                     791609    13.0         53976 
31 New Mexico     McKinley                   73998    15.5         28772 
32 New York       Bronx                    1428357    14.0         34299 
33 North Carolina Anson                      26135    17.6         33228 
34 North Dakota   Sioux                       4380    20.3         38068 
35 Ohio           Adams                      28229    13.4         35560 
36 Ohio           Pike                       28396    13.4         40283 
37 Oklahoma       McIntosh                   20280    10.8         36011 
38 Oregon         Harney                      7229    16.1         37580 
39 Pennsylvania   Philadelphia             1555072    13.9         38253 
40 Rhode Island   Providence                630459     9.5         49743 
41 South Carolina Allendale                   9838    22.6         25327 
42 South Dakota   Corson                      4149    29.4         31676 
43 Tennessee      Hardeman                   26253    18.7         31801 
44 Texas          Brooks                      7221    19.8         22741 
45 Utah           Iron                       47139    10.5         43855 
46 Vermont        Essex                       6207     9.4         36599 
47 Virginia       Emporia city                5672    19.1         28601 
48 Washington     Grays Harbor               71419    14.3         43538 
49 West Virginia  Calhoun                     7557    18.1         35568 
50 Wisconsin      Menominee                   4451    14.8         35343 
51 Wyoming        Fremont                    40755     7.8         52773

### Number of observations

In [49]:
counties_selected %>%
    group_by(state) %>%
    top_n(3, unemployment)

state         county                    population unemployment income
1   Alabama       Conecuh                    12865     22.6         24900 
2   Alabama       Monroe                     22217     20.7         27257 
3   Alabama       Wilcox                     11235     20.8         23750 
4   Alaska        Bethel Census Area         17776     17.6         51012 
5   Alaska        Northwest Arctic Borough    7732     21.9         63648 
6   Alaska        Yukon-Koyukuk Census Area   5644     18.2         38491 
7   Arizona       Apache                     72124     18.2         31757 
8   Arizona       Graham                     37407     14.1         45964 
9   Arizona       Navajo                    107656     19.8         35921 
10  Arkansas      Desha                      12379     17.7         27197 
11  Arkansas      Lee                         9966     17.3         28393 
12  Arkansas      Phillips                   20391     18.1         26844 
13  California    Imperial                  178206     17.4         41079 
14  California    Merced                    263885     16.2         42462 
15  California    Plumas                     18966     15.2         47333 
16  California    Stanislaus                527367     15.2         50125 
17  Colorado      Costilla                    3581     14.0         31321 
18  Colorado      Crowley                     5551     27.0         31151 
19  Colorado      Custer                      4303     14.0         35000 
20  Colorado      Huerfano                    6502     15.0         31715 
21  Connecticut   Hartford                  896943      9.3         66395 
22  Connecticut   New Haven                 862224      9.5         61640 
23  Connecticut   Windham                   117470      9.3         59392 
24  Delaware      Kent                      169509      8.4         54976 
25  Delaware      New Castle                549643      7.4         65476 
26  Delaware      Sussex                    207302      8.0         53751 
27  Florida       Columbia                   67806     14.8         41926 
28  Florida       Hamilton                   14395     15.8         35048 
29  Florida       Holmes                     19635     15.4         35020 
30  Georgia       Clay                        3104     19.2         20438 
... ...           ...                       ...        ...          ...   
124 South Dakota  Corson                     4149      29.4         31676 
125 South Dakota  Todd                       9942      25.4         31128 
126 South Dakota  Ziebach                    2833      27.4         35119 
127 Tennessee     Hardeman                  26253      18.7         31801 
128 Tennessee     McNairy                   26103      14.9         32557 
129 Tennessee     Meigs                     11716      15.0         34424 
130 Texas         Brooks                     7221      19.8         22741 
131 Texas         Jim Hogg                   5239      16.5         35521 
132 Texas         San Augustine              8695      16.3         28275 
133 Utah          Garfield                   5069       8.6         42614 
134 Utah          Iron                      47139      10.5         43855 
135 Utah          San Juan                  15152       8.4         41484 
136 Vermont       Caledonia                 31012       7.2         45323 
137 Vermont       Essex                      6207       9.4         36599 
138 Vermont       Grand Isle                 6952       7.4         62608 
139 Virginia      Emporia city               5672      19.1         28601 
140 Virginia      Franklin city              8457      13.7         32399 
141 Virginia      Petersburg city           32123      13.0         31798 
142 Washington    Ferry                      7652      12.2         38125 
143 Washington    Grays Harbor              71419      14.3         43538 
144 Washington    Mason                     60791      12.9         50406 
145 West Virginia Braxton    

`top_n` is often used when creating graphs.

## Exercise

In [50]:
# Selecting a county from each region

counties_selected <- counties %>%
  select(region, state, county, metro, population, walk)

counties_selected %>%
  # Group by region
  group_by(region) %>%
  # Find the greatest number of citizens who walk to work
  top_n(1, walk)

region        state        county                 metro    population walk
1 West          Alaska       Aleutians East Borough Nonmetro    3304    71.2
2 Northeast     New York     New York               Metro    1629507    20.7
3 North Central North Dakota McIntosh               Nonmetro    2759    17.5
4 South         Virginia     Lexington city         Nonmetro    7071    31.7

*Notice that three of the places lots of people walk to work are low-population nonmetro counties, but that New York City also pops up.*

In [53]:
# Finding the highest-income state in each region

counties_selected <- counties %>%
  select(region, state, county, population, income)

counties_selected %>%
  group_by(region, state) %>%
  # Calculate average income
  summarise(average_income = mean(income)) %>%
  # Find the highest income state in each region
  top_n(1, average_income)

`summarise()` has grouped output by 'region'. You can override using the `.groups` argument.


region        state        average_income
1 North Central North Dakota 55574.87      
2 Northeast     New Jersey   73014.10      
3 South         Maryland     69200.38      
4 West          Alaska       65124.54

*New Jersey in the Northeast is the state with the highest* `average_income` *of 73014.*

In [55]:
# Using summarize, top_n, and count together

counties_selected <- counties %>%
  select(state, metro, population)

counties_selected %>%
  # Find the total population for each combination of state and metro
  group_by(state, metro) %>%
  summarise(total_pop = sum(population)) %>%
  # Extract the most populated row for each state
  top_n(1, total_pop)

`summarise()` has grouped output by 'state'. You can override using the `.groups` argument.


state          metro    total_pop
1  Alabama        Metro     3671377 
2  Alaska         Metro      494990 
3  Arizona        Metro     6295145 
4  Arkansas       Metro     1806867 
5  California     Metro    37587429 
6  Colorado       Metro     4590896 
7  Connecticut    Metro     3406918 
8  Delaware       Metro      926454 
9  Florida        Metro    18941821 
10 Georgia        Metro     8233886 
11 Hawaii         Metro     1145126 
12 Idaho          Metro     1069978 
13 Illinois       Metro    11368647 
14 Indiana        Metro     5095864 
15 Iowa           Metro     1804496 
16 Kansas         Metro     1941587 
17 Kentucky       Metro     2558548 
18 Louisiana      Metro     3856617 
19 Maine          Metro      781723 
20 Maryland       Metro     5777808 
21 Massachusetts  Metro     6606838 
22 Michigan       Metro     8095885 
23 Minnesota      Metro     4180337 
24 Mississippi    Nonmetro  1628768 
25 Missouri       Metro     4490148 
26 Montana        Nonmetro   655927 
27 Nebraska       Metro     1199134 
28 Nevada         Metro     2529002 
29 New Hampshire  Metro      828251 
30 New Jersey     Metro     8904413 
31 New Mexico     Metro     1388935 
32 New York       Metro    18267314 
33 North Carolina Metro     7639992 
34 North Dakota   Nonmetro   366754 
35 Ohio           Metro     9198923 
36 Oklahoma       Metro     2504090 
37 Oregon         Metro     3286017 
38 Pennsylvania   Metro    11285024 
39 Rhode Island   Metro     1053661 
40 South Carolina Metro     4023755 
41 South Dakota   Nonmetro   430604 
42 Tennessee      Metro     5007264 
43 Texas          Metro    23510680 
44 Utah           Metro     2591224 
45 Vermont        Nonmetro   411523 
46 Virginia       Metro     7195353 
47 Washington     Metro     6272805 
48 West Virginia  Metro     1137523 
49 Wisconsin      Metro     4237349 
50 Wyoming        Nonmetro   404237

In [56]:
counties_selected %>%
  # Find the total population for each combination of state and metro
  group_by(state, metro) %>%
  summarize(total_pop = sum(population)) %>%
  # Extract the most populated row for each state
  top_n(1, total_pop) %>%
  # Count the states with more people in Metro or Nonmetro areas
  ungroup() %>%
  count(metro)

`summarise()` has grouped output by 'state'. You can override using the `.groups` argument.


metro    n 
1 Metro    44
2 Nonmetro  6

*Notice that 44 states have more people living in Metro areas, and 6 states have more people living in Nonmetro areas.*